### ISSUE TITLE
Proposal to enhance defining input/output variables.

### ISSUE THREAD

**Issue by dj2** [gpuweb/gpuweb#1155]
Writing multiple shaders in a single file could benefit from collapsing, or moving module scoped input/output variables.

There are a few different possibilities here depending on how we want to proceed.

# Multiple storage classes
```rust
[[location(0)]] var<out(vertex), in(fragment)> frag_color : vec4<f32>
```

# Multiple locations
```rust
[[location(0, fragment), location(1, vertex)]] var<in(vertex), out(fragment)> colour : vec4<f32>
```

# As structs
```rust
struct PerVertex {
  [[location(0)]] colour: vec4<f32>;
};
fn vertex() -> PerVertex {... }
fn fragment(v: PerVertex) -> vec4<f32> { .. }
```

# As entry point parameters
```rust
fn fragment(
    [[location(0)]] colour : vec4<f32>,
) -> [[location(1)]] vec4<f32> {
}

[[stage(fragment)]]
fn frag(
  [[buildin(position)]] pos : vec4<f32>) -> void {}

[[stage(compute)]]
fn comp_main(
  [[builtin(invocation_id)]] gl_GlobalInvocationId : vec3<u32>,
  [[set(0), binding(0), uniform]] params : SimParams,
  [[set(0), binding(1), storage]] particlesA : Particles,
  [[set(0), binding(2), storage]] particlesB : Particles) -> void {}
```

**Comment by litherum** [comment]
Another possibility:

```
[[stage(vertex)]] fn v(...) -> [[location(0)]] colour : vec4<f32>, [[location(1)]] vec4<f32>
```

**Comment by dneto0** [comment]
Discussed in the virtual [F2F 2020-10-23](https://docs.google.com/document/d/1p1jREMgonZszFJ8eastnnDb36rKKm7xnDYnMjC6zj5c/edit#heading=h.zg7fj1vgmcj7) 

I had sketched something similar to what @litherum  showed above, but I had parentheses:

```
[[stage(vertex)]] fn myvertexShader(
[[location(0)]] a: i32, 
[[location(1)]] b:f32) 
-> ( [[location(0)]] outa: i32, [[location(1)]] outb: f32) {
    outa = 2*a;
    outb = 2*b; 
}
```
Because this is an entry point, outa and outb are as if declared as var<out>. So the are default-zero-initialized and we can read and write to them.

If we think about extending this to helper functions, then after-arrow declarations are … an open question because now they might be the RHS of an assignment in the caller.  Then the next obvious step is to do parallel tuple-assignment.  (But you have to be careful about specifying just enough ordering to remain obvious and safe.)

**Comment by litherum** [comment]
Right.

```
const a : f32, b : f32 = myHelper();
```

**Comment by dneto0** [comment]
But weirdly, the tuple-after-arrow syntax doesn't save any typing, unless you also predeclare a struct that will have the extra noise for repeating all the members, and the location assignments?  I would still need to see more worked examples.

**Comment by dneto0** [comment]
(Post-MVP I'd like to eventually get to type inference that will let us shorten to something like:

a, b := myHelper();  // for const, non-pointer result.
var a, b := myHelper(); // for var decl initializations
)

**Comment by litherum** [comment]
There's another possibility that didn't get listed above:

```
struct Intermediate {
    [[location(0)]] colour : vec4<f32>;
    [[location(1)]] foobar : vec4<f32>;
}

[[stage(vertex)]] fn v(...) -> Intermediate {
    var result = Intermediate();
    result.colour = ...; // Pack the struct
    ...
    return result;
}

[[stage(fragment)]] fn f(Intermediate intermediate) -> ... {
    let unpacked : vec4<f32> = intermediate.colour; // Unpack the struct
    ...
}

```

**Comment by kainino0x** [comment]
When designing this it's worth noting that sometimes a vertex shader will be paired with a fragment shader that consumes only a subset of the vertex outputs.

**Comment by litherum** [comment]
In the call we were discussing something like this:

struct VertexOutput {
    [[location(0)]] colour : vec4<f32>;
    [[location(1)]] foobar : vec4<f32>;
}

[[stage(vertex)]] fn v(...) -> VertexOutput {
    var result : VertexOutput = VertexOutput();
    result.colour = ...; // Pack the struct
    ...
    return result;
}

struct PartialFragmentInput {
    [[location(1)]] foobar : vec4<f32>;
}

[[stage(fragment)]] fn f(partialFragmentInput : PartialFragmentInput, [[location(0)]] colour : vec4<f32>) -> ... {
    let unpacked : vec4<f32> = partialFragmentInput.foobar; // Unpack the struct
    let unpacked2 : vec4<f32> = colour; // But you don't have to unpack everything
    ...
}

**Comment by dneto0** [comment]
In the options where the inputs/outputs are function parameters and returns, how are built-in variables handled?
Right now we can do this:

```
[[builtin(position)]] var<out> pos : vec4<f32>;

fn bar()  {  pos = vec4<f32>(); } // accesses the builtin variable directly.  It knows it's a builtin.

[[stage(vertex)]] fn entry() -> void {  bar(); }
```

In the new intputs/outputs are function parameters, you presumably have to pass pointer/references around:

```


fn bar( posref : ptr<out,vec4<f32>> )  {  
   posref = vec4<f32>();  // accesses the builtin variable directly.  It knows it's a builtin.
} 


[[stage(vertex)]] fn entry([[builtin(position)]] pos : vec4<f32>; ) -> void { 
    bar(pos); // where the evaluation of pos is as pointer-to-output-vec4<f32>
}
```
Doesn't that make accessing builtins really quite ugly in helper functions?

I would guess that mapping to HLSL would require the `inout` modifier? (Maybe that's already baked in with ptr parameters.)

**Comment by magcius** [comment]
From my experience working on multiple engines, the struct is honestly the nicest approach, since otherwise you basically end up writing your own struct and then packing/unpacking at the beginning/end anyway. Structs are really convenient for parameter passing.

**Comment by dneto0** [comment]
Another thing to handle:  You can have an  initializer for an output variable, but you can't have one for an input variable.

**Comment by ben-clayton** [comment]
A post-MVP feature request - passing multiple-return values to another function without intermediates:

Given:

```rust
fn fetch_gbuffer() -> (normal : vec3<f32>, diffuse : vec3<f32>)  { ... }
fn calc_lighting(ambient : vec3<f32>, normal : vec3<f32>, diffuse : vec3<f32>, specular f32) { ... }
```

It would be very convenient to be able to call `calc_lighting()` like:

```rust
calc_lighting(ambient, fetch_gbuffer(), specular);
```

**Comment by jdashg** [comment]
I really like being able to re-use structs for vert-out/frag-in. This is the primary (though not only) way in HLSL, and it's great for unambiguously sharing the interface on both sides of rasterization interpolation.

**Comment by kvark** [comment]
I tried to describe the arguments/result types proposal in #1315

**Comment by kvark** [comment]
Closed by #1426
